tests/testcafe/tests.js

```diff
@@ -300,18 +275,62 @@ test('Dynamic expressions', async t => {
     .click('[id="jyo6"]')
   const log = await Selector('#jeLog').textContent
   for (let i=0; i<ops.length; i++) {
-    await t.expect(log).contains('"'+ops[i][1]+'": '+ops[i][2])
+    const logContains = log.includes('"'+ops[i][1]+'": '+ops[i][2]);
+    await t.expect(logContains)
+           .ok('Test "' + ops[i] + '" failed.');
   };
   await t
     .pressKey('ctrl+j')
 });
 
+function publicLibraryTest(game, variant, md5, tests) {
+  test.after(async t => {
+    await removeGame(t);
+    await t.expect(Selector('#statesOverlay').visible).ok();
+  })(`Public library: ${game} (variant ${variant})`, async t => {
+    await ClientFunction(prepareClient)();
+    await t
+      .pressKey('esc')
+      .click('#statesButton')
+      .click(Selector('td.name').withExactText(game).prevSibling().child())
+      .hover('.roomState')
+      .click(Selector('button.play').nth(variant));
+    await setName(t); 
+    await tests(t); 
+    await compareState(t, md5);
+  });
+}
+
+function publicLibraryButtons(game, variant, md5, tests) {
+  publicLibraryTest(game, variant, md5, async t => {
+      for(const b of tests)
+        if(typeof b == "string") {
+          await t.click(`[id="${b}"]`)
+        } else {
+          await t.dragToElement(b[0](), b[1](), { speed:0.5 });
+        }
+  });
+}
+
 publicLibraryButtons('Blue',               0, '096bdf3bd07bb277c2c1f4cc132f0695', [
   'fcc3fa2c-c091-41bc-8737-54d8b9d3a929', 'd3ab9f5f-daa4-4d81-8004-50a9c90af88e_incrementButton',
   'd3ab9f5f-daa4-4d81-8004-50a9c90af88e_incrementButton', 'd3ab9f5f-daa4-4d81-8004-50a9c90af88e_decrementButton',
   'reset_button', 'buttonInputGo', 'fcc3fa2c-c091-41bc-8737-54d8b9d3a929', '9n2q'
 ]);
+publicLibraryButtons('Dots',               0, '23894df38f786cb014fa1cd79f2345db', [ 'reset', 'buttonInputGo', 'col11', 'col21', 'col12', 'col22', 'row11', 'row31', 'row21', 'row32', 'row12', 'row42', 'row22', 'row23', 'col23' ]);
+publicLibraryButtons('Reversi',            0, '35e0017570f9ecd206a2317c1528be36',
+       [
+         [ ()=>{return Selector("#zpiece15")}, ()=>{return Selector("#sq23")} ],
```

Comment:
`()=>{return Selector("#zpiece15")}` should be the same as `()=>Selector("#zpiece15")` (note the missing braces). Just FYI.

Reply:
Thanks. Good to know. Fixed.